src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -540,45 +553,38 @@ private async Task RefreshKeyValueCollections(CancellationToken cancellationToke
-                        var keyPrefix = changeWatcher.Key.Substring(0, changeWatcher.Key.Length - 1);
-                        currentKeyValues = _applicationSettings.Values.Where(kv =>
-                        {
-                            return kv.Key.StartsWith(keyPrefix) && kv.Label == changeWatcher.Label.NormalizeNull();
-                        });
-                    }
-                    else
+                if (changeWatcher.Key.EndsWith("*"))
+                {
+                    // Get current application settings starting with changeWatcher.Key, excluding the last * character
+                    var keyPrefix = changeWatcher.Key.Substring(0, changeWatcher.Key.Length - 1);
+                    currentKeyValues = _applicationSettings.Values.Where(kv =>
                     {
-                        currentKeyValues = _applicationSettings.Values.Where(kv =>
-                        {
-                            return kv.Key.Equals(changeWatcher.Key) && kv.Label == changeWatcher.Label.NormalizeNull();
-                        });
-                    }
-
-                    keyValueChanges = await _client.GetKeyValueChangeCollection(
-                        currentKeyValues,
-                        new GetKeyValueChangeCollectionOptions
-                        {
-                            KeyFilter = changeWatcher.Key,
-                            Label = changeWatcher.Label.NormalizeNull(),
-                            RequestTracingEnabled = _requestTracingEnabled,
-                            RequestTracingOptions = _requestTracingOptions
-                        },
-                        cancellationToken).ConfigureAwait(false);
-
-                    success = true;
+                        return kv.Key.StartsWith(keyPrefix) && kv.Label == changeWatcher.Label.NormalizeNull();
+                    });
                 }
-                finally
+                else
                 {
-                    UpdateCacheExpirationTime(changeWatcher, success);
+                    currentKeyValues = _applicationSettings.Values.Where(kv =>
+                    {
+                        return kv.Key.Equals(changeWatcher.Key) && kv.Label == changeWatcher.Label.NormalizeNull();
+                    });
                 }
 
+                keyValueChanges = await client.GetKeyValueChangeCollection(
+                    currentKeyValues,
+                    new GetKeyValueChangeCollectionOptions
+                    {
+                        KeyFilter = changeWatcher.Key,
+                        Label = changeWatcher.Label.NormalizeNull(),
+                        RequestTracingEnabled = _requestTracingEnabled,
+                        RequestTracingOptions = _requestTracingOptions
+                    },
+                    cancellationToken).ConfigureAwait(false);
+
+                UpdateCacheExpirationTime(changeWatcher, success: true);

```

Comment:
I would still keep `UpdateCacheExpirationTime` in the `finally` clause just to make sure it's called (even if you just call it in the success case). It's so easy to miss it out in current code if we make any future changes.

Reply:
With the updated logic, `UpdateCacheExpirationTime` gets called in finally block.